glean-core/android/src/main/java/mozilla/telemetry/glean/scheduler/MetricsPingScheduler.kt

```diff
@@ -254,9 +274,27 @@ internal class MetricsPingScheduler(
      * @param now a [Calendar] instance representing the current time.
      */
     @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
-    internal fun collectPingAndReschedule(now: Calendar) {
-        Log.i(LOG_TAG, "Collecting the 'metrics' ping, now = ${now.time}")
-        Pings.metrics.submit()
+    internal fun collectPingAndReschedule(now: Calendar, startupPing: Boolean = false) {
+        Log.i(LOG_TAG, "Collecting the 'metrics' ping, now = ${now.time}, startup = $startupPing")
+        if (startupPing) {
+            // **IMPORTANT**
+            //
+            // During the Glean initialization, we require any metric recording to be
+            // batched up and replayed after any startup metrics ping is sent. To guarantee
+            // that, we dispatch this function from `Dispatchers.API.executaTask`. However,
+            // Pings.metrics.submit() ends up calling `Dispatchers.API.launch` again which
+            // will delay the ping collection task after any pending metric recording is
+            // executed, breaking the 'metrics' ping promise of sending a startup 'metrics'
+            // ping only containing data from the previous session.
+            // To prevent that, we synchronously manually dispatch the 'metrics' ping, without
+            // going through our public API.
+            //
+            // * Do not change this line without checking what it implies for the above wall
+            // of text. *
```

Comment:
Sounds like something a test should catch

Reply:
Yes, the provided test catches that. However I think it's still beneficial to call this out explicitly in the comment, so that we don't end up scratching our heads on "why is the test failing" months from now :)